File: Data Structures & Algorithms/count-number-of-islands/submission-5.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        
        if not grid:
            return 0

        count = 0

        rows = len(grid)
        cols = len(grid[0])

        def dbf(r,c):

            if r < 0 or r >= rows or c < 0 or c >= cols or grid[r][c] == '0':

                return

            grid[r][c] = '0'

            dbf(r+1,c)
            dbf(r-1,c)
            dbf(r,c+1)
            dbf(r,c-1)

        for r in range(rows):

            for c in range(cols):

                if grid[r][c] == '1':

                    count = count + 1

                    dbf(r,c)

        return count

File: Data Structures & Algorithms/max-area-of-island/submission-5.java
class Solution {
    public int maxAreaOfIsland(int[][] grid) {
        
        int rows = grid.length;

        int cols = grid[0].length;

        int max_area = 0;

        for(int i; i < rows; i++){

            for(int j; j < cols; j++){

                if(grid[i][j] == 1){

                    area(grid,i,j);

                    max_area = Math.max(max_area,area(grid,i,j));
                }
            }
        }
        return max_area;
    }

    private int area(int[][] grid,int r,int c){

        int rows = grid.length;

        int cols = grid[0].length;

        if(r < 0 || r >= rows || c < 0 || c >= cols){

            return 0;
        }

        if(grid[r][c] == 0) return 0;

        grid[r][c] = 0;

        return 1 + area(r+1,c) + area(r-1,c) + area(r,c+1) + area(r,c-1);
    }
}
